benchmark/src/DriverBench/ParallelBench.php

```diff
@@ -0,0 +1,210 @@
+            // Wait for a child process to finish if we have reached the maximum number of processes
+            if (count($pids) >= $params['processes']) {
+                $pid = pcntl_waitpid(-1, $status);
+                unset($pids[$pid]);
+            }
+
+            $pid = pcntl_fork();
+            if ($pid === 0) {
+                // Reset to ensure that the existing libmongoc client (via the Manager) is not re-used by the child
+                // process. When the child process constructs a new Manager, the differing PID will result in creation
+                // of a new libmongoc client.
+                Utils::reset();
+                self::importFile($file);
+
+                // Exit the child process
+                exit(0);
+            }
+
+            if ($pid === -1) {
+                throw new RuntimeException('Failed to fork');
+            }
+
+            // Keep the forked process id to wait for it later
+            $pids[$pid] = true;
+        }
+
+        // Wait for all child processes to finish
+        while ($pids !== []) {
+            $pid = pcntl_waitpid(-1, $status);
+            unset($pids[$pid]);
+        }
+    }
+
+    /**
+     * Parallel: LDJSON multi-file import
+     * Using amphp/parallel-functions with worker pool
+     *
+     * @param array{processes:int, files:string[], batchSize:int} $params
+     */
+    #[BeforeMethods('beforeMultiFileImport')]
+    #[ParamProviders(['provideProcessesParameter'])]
+    #[Revs(1)]
+    #[Iterations(1)]
+    public function benchMultiFileImportAmp(array $params): void
+    {
+        wait(parallelMap(
+            self::getFileNames(),
+            // Uses array callable instead of closure to skip complex serialization
+            [self::class, 'importFile'],
+            // The pool size is the number of processes
+            new DefaultPool($params['processes']),
+        ));
+    }
+
+    public static function provideProcessesParameter(): Generator
+    {
+        yield '1 proc' => ['processes' => 1]; // 100 sequences, to compare to the single thread baseline
+        yield '2 proc' => ['processes' => 2]; // 50 sequences
+        yield '4 proc' => ['processes' => 4]; // 25 sequences
+        yield '8 proc' => ['processes' => 8]; // 13 sequences
+        yield '13 proc' => ['processes' => 13]; // 8 sequences
+        yield '20 proc' => ['processes' => 20]; // 5 sequences
+        yield '34 proc' => ['processes' => 34]; // 3 sequences
+    }
+
+    public function beforeMultiFileImport(): void
+    {
+        $database = Utils::getDatabase();
+        $database->drop();
+        $database->createCollection(Utils::getCollectionName());
+    }
+
+    public static function importFile(string $file): void
+    {
+        $namespace = sprintf('%s.%s', Utils::getDatabaseName(), Utils::getCollectionName());
+
+        $bulkWrite = new BulkWrite();
+        foreach (file($file, FILE_IGNORE_NEW_LINES | FILE_NO_DEFAULT_CONTEXT) as $line) {
```

Comment:
I see `file()` used here and above in `benchMultiFileImportInsertMany()`. Is there any memory concern with loading the entire file into an array?

Asking as I'm not sure how large these are, and if we're also parsing them into BSON then we're likely storing multiple copies of the data in memory. If so, I suppose we could optimize the file reading to use something like [SplFileObject](https://www.php.net/SplFileObject) instead to stream the data.

I see this function is also used _within_ some benchmarks (e.g. `benchMultiFileImportAmp`). Is this something we should be doing in a setup step instead, or does the benchmark actually want to time reading the file from disk?

If the file reading is intentionally part of the benchmark, then maybe you can compare `file()` and `SplFileObject` and use whichever is more performant.

Feel free to resolve if this isn't an issue.